### PR TITLE
Zoe simpleExchange perf test in swingset-runner

### PR DESCRIPTION
**Pull request by FUDCo**
Also a port of the Zoe swingset test to swingset-runner.

### CODE REVIEW

**Review by Chris-Hibbert — APPROVED** [review]

**Comment by warner** [comment]
Is any of this covered by unit tests?

I'm worried that a lot of this code is super-sensitive to how Zoe works internally, but if it doesn't live in the Zoe package, it won't get updated when Zoe does. And without any test coverage, it'll probably break. I know that personally, my new-SES branch changes some of these setup patterns, and I rely upon the unit tests to tell me what I need to update.

Maybe it's a question fo @Chris-Hibbert , but is there a way we could have most of these new files live in `packages/zoe/` or `packages/ERTP/` instead? I'd love it if the same vats could be used for both `zoe/tests/swingsetTests/` and for performance work. That might involve a few extra methods in a bootstrap vat, so there's `bootstrap` that's mostly for the benefit of the unit tests, and `runBenchmarkRound` for perf.. maybe we need to fuss with the `argv` mode-switcher code, or maybe we should change the zoe tests to `c.queueToExport` a second (non-`bootstrap`) method invocation to tell it which test to run.

Then `swingset-runner` unit tests could include (short) perf runs, which import the vats from zoe+ERTP, but run them with an `argv` mode (or `queueToExport` some "prepare for perf run" methods) that tells them to do perf-ish things instead of unit-test-ish things.

**Comment by Chris-Hibbert** [comment]
>  is there a way we could have most of these new files live in packages/zoe/ or packages/ERTP/ instead? 

It's a good question, but I'd start by asking @FUDCo how much he had to modify to get here, and why he wasn't able to share the code from packages/zoe. If we can make it work, it would definitely be a win to have only one copy of these tests, and run them for multiple tests.

File: packages/swingset-runner/demo/exchangeBenchmark/bootstrap.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import harden from '@agoric/harden';
+
+import produceIssuer from '@agoric/ertp';
+import { makePrintLog } from './printLog';
+
+/* eslint-disable import/no-unresolved, import/extensions */
+import simpleExchangeBundle from './bundle-simpleExchange';
+/* eslint-enable import/no-unresolved, import/extensions */
```
[the remark this answers is not in the document]
Comment (FUDCo) [reply packages/swingset-runner/demo/exchangeBenchmark/bootstrap.js]: I don't actually know.  This part was cribbed from the Zoe tests directly.

File: packages/swingset-runner/demo/zoeTests/vat-alice.js [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+      Contribution1: moolaIssuer,
+      Contribution2: simoleanIssuer,
+    });
+    const refundInvite = await E(zoe).makeInstance(
+      installId,
+      issuerKeywordRecord,
+    );
+
+    const instanceHandle = await getInstanceHandle(refundInvite);
+    const instanceRecord = await E(zoe).getInstanceRecord(instanceHandle);
```
[the remark this answers is not in the document]
Comment (FUDCo) [reply packages/swingset-runner/demo/zoeTests/vat-alice.js:27]: Once again, another case of what Brian calls cargo-cult-and-paste.  This is how the Zoe tests did it and I just replicated what was there.  But it's a good point.  I'll take another pass and look at that.

**Comment by FUDCo** [comment]
> > is there a way we could have most of these new files live in packages/zoe/ or packages/ERTP/ instead?
> 
> It's a good question, but I'd start by asking @FUDCo how much he had to modify to get here, and why he wasn't able to share the code from packages/zoe. If we can make it work, it would definitely be a win to have only one copy of these tests, and run them for multiple tests.

Much of the answer to your question is written up in #1194

**Comment by FUDCo** [comment]
> I'm worried that a lot of this code is super-sensitive to how Zoe works internally, but if it doesn't live in the Zoe package, it won't get updated when Zoe does. And without any test coverage, it'll probably break. I know that personally, my new-SES branch changes some of these setup patterns, and I rely upon the unit tests to tell me what I need to update.

This should not be too entangled with Zoe internals, since this is written to the Zoe APIs.  The one exception is the import of `@agoric/zoe/src/clientSupport`, which does concern me a little but it's trivial helper code rather than deep mechanism and could either be added to the the official exports of Zoe or just copied.  The thing that gives me pause was the need to implement the `prepareContracts.js` utility, which generates the contract bundle and would need to be rerun if the definitions of any of the contracts change.  The Zoe swingset test avoids this by doing this in the test driver (indeed, the guts of `prepareContracts` are stolen directly from there).  I would be happier if there was an officially maintained installable bundle that could be used directly.

**Comment by Chris-Hibbert** [comment]
Here's a pretty simple idea for maintaining compatibility: Have the sources of the modified version for performance tests live in Zoe's repo with the repetition count set to 1. This will serve as a redundant set of tests, that can be imported by the testing framework to run multiple times, and we'll maintain them as we update Zoe's APIs. It does mean that the entire suite of tests runs more slowly, but I think that could be worth it.

File: packages/swingset-runner/demo/exchangeBenchmark/bootstrap.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (FUDCo) [reply packages/swingset-runner/demo/exchangeBenchmark/bootstrap.js]: Mystery solved: original Zoe tests had a whole block of imports here (as does my `zoeTests` test.

**Review by warner — COMMENTED** [review]
You landed first, so it's my responsibility to adjust my (new-ses) PR to what's on trunk now. But since this new code isn't run by any tests, I'll need to pull you into reviewing my branch to make sure I didn't break it. I'm guessing you've got some recipe for manually running all the demos; you'll need to run that and let me know it it worked.

I'd really like it if this new code were exercised somehow by `yarn test`. Be prepared to get tagged for reviews of anything swingset-ish or zoe-ish that might possibly break it until you get tired of manually testing the runner and add some automation :).

File: packages/swingset-runner/package.json [hunk 3]
```diff
@@ -14,12 +14,61 @@
...
+    "@agoric/ertp": "^0.5.5",
+    "@agoric/eventual-send": "^0.9.2",
     "@agoric/harden": "^0.0.8",
+    "@agoric/same-structure": "^0.0.7",
     "@agoric/stat-logger": "^0.2.5",
     "@agoric/swing-store-lmdb": "^0.3.2",
     "@agoric/swing-store-simple": "^0.2.2",
-    "@agoric/swingset-vat": "^0.5.2"
+    "@agoric/swingset-vat": "^0.5.2",
+    "@agoric/zoe": "^0.6.2"
```
[the remark this answers is not in the document]
Comment (FUDCo) [reply packages/swingset-runner/package.json:27]: I was conflicted over these also; they're in there strictly to make eslint happy but they really don't belong.  And they're not really devDepencies either.  They want to be a third kind of dependency, namely for the demos themselves.  I don't know if this is an affordance that's available in `package.json` or not - or if maybe there should be a lower level `package.json` file just for the demos (or maybe one for each as needed) but I'm not sure how works with respect to packaging.  Who's our local `package.json` guru?

**Pull request merged**